integrationtests/agent/adoption_test.go

```diff
@@ -0,0 +1,387 @@
+					DefaultNamespace: namespace,
+					Helm: &v1alpha1.HelmOptions{
+						TakeOwnership: takeOwnership,
+					},
+				},
+			},
+		}
+
+		err := k8sClient.Create(context.TODO(), &bundled)
+		Expect(err).To(BeNil())
+		Expect(bundled).To(Not(BeNil()))
+		Expect(bundled.Spec.DeploymentID).ToNot(Equal(bundled.Status.AppliedDeploymentID))
+		Expect(bundled.Status.Ready).To(BeFalse())
+		Eventually(func() bool {
+			err := k8sClient.Get(context.TODO(), types.NamespacedName{Namespace: clusterNS, Name: name}, &bundled)
+			if err != nil {
+				return false
+			}
+			return bundled.Status.Ready
+		}).Should(BeTrue(), "BundleDeployment not ready: status: %+v", bundled.Status)
+		Expect(bundled.Spec.DeploymentID).To(Equal(bundled.Status.AppliedDeploymentID))
+	}
+
+	waitForConfigMap := func(name string) {
+		Eventually(func() error {
+			_, err := env.getConfigMap(name)
+			return err
+		}).Should(Succeed())
+	}
+
+	createConfigMap := func(data, labels, annotations map[string]string) *corev1.ConfigMap {
+		cm := &corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:        "cm1",
+				Namespace:   namespace,
+				Labels:      labels,
+				Annotations: annotations,
+			},
+			Data: data,
+		}
+		Expect(k8sClient.Create(ctx, cm)).To(Succeed())
+		waitForConfigMap("cm1")
+		return cm
+	}
+
+	// assertConfigMap checks that the ConfigMap exists and that it passes the
+	// provided validate function.
+	assertConfigMap := func(validate func(corev1.ConfigMap) error) {
+		cm := corev1.ConfigMap{}
+		var err error
+		Eventually(func() error {
+			err = k8sClient.Get(
+				ctx,
+				types.NamespacedName{Namespace: namespace, Name: "cm1"},
+				&cm,
+			)
+			if err != nil {
+				return err
+			}
+			return validate(cm)
+		}).Should(Succeed(), "assertConfigMap error: %v in %+v", err, cm)
+	}
+
+	// assertBundleDeployment checks that the BundleDeployment exists and that it
+	// passes the provided validate function.
+	assertBundleDeployment := func(name string, validate func(*v1alpha1.BundleDeployment) error) {
+		bd := v1alpha1.BundleDeployment{}
+		var err error
+		Eventually(func() error {
+			err = k8sClient.Get(
+				ctx,
+				types.NamespacedName{Namespace: clusterNS, Name: name},
+				&bd,
+			)
+			if err != nil {
+				return err
+			}
+			return validate(&bd)
+		}).Should(Succeed(), "assertBundleDeployment: error %v in %+v", err, bd)
+	}
+
+	// mapPartialMatch checks that the super map contains all the keys and values
+	// of the sub map. If the value in the sub map is an empty string, the key
+	// must exist in the super map but the value is not compared.
+	mapPartialMatch := func(super, sub map[string]string) error {
+		for k, v := range sub {
+			if v == "" {
+				if _, ok := super[k]; !ok {
+					return fmt.Errorf("key %s not found in %+v", k, super)
+				}
+			} else {
+				if v2, ok := super[k]; !ok || v2 != v {
+					return fmt.Errorf("key %s not found or value %s does not match %s in %+v", k, v2, v, super)
+				}
+			}
```

Comment:
I was kind of happy with the mapPartialMatch function, but thinking about it, it has benefits to use Expect rather than returning values and having to handle them (in tests). So, I've rewritten it again and the result is a really nice representation of test failures. I hope you will ike it, too.